src/jace/stages.py

```diff
@@ -0,0 +1,333 @@
+        jit_options: Mapping[str, Any],
+    ) -> None:
+        super().__init__()
+        # We have to shallow copy both the translator and the jit options.
+        #  This prevents that any modifications affect `self`.
+        #  Shallow is enough since the translators themselves are immutable.
+        self._primitive_translators = dict(primitive_translators)
+        # TODO(phimuell): Do we need to deepcopy the options?
+        self._jit_options = dict(jit_options)
+        self._fun = fun
+
+    def __call__(
+        self,
+        *args: Any,
+        **kwargs: Any,
+    ) -> Any:
+        """Executes the wrapped function, lowering and compiling as needed in one step.
+
+        The arguments passed to this function are the same as the wrapped function uses.
+        """
+
+        # If we are inside a traced context, then we forward the call to the wrapped function.
+        #  This ensures that JaCe is composable with Jax.
+        if util.is_tracing_ongoing(*args, **kwargs):
+            return self._fun(*args, **kwargs)
+
+        lowered = self.lower(*args, **kwargs)
+        compiled = lowered.compile()
+        return compiled(*args, **kwargs)
+
+    @tcache.cached_transition
+    def lower(
+        self,
+        *args: Any,
+        **kwargs: Any,
+    ) -> JaCeLowered:
+        """Lower this function explicitly for the given arguments.
+
+        Performs the first two steps of the AOT steps described above, i.e. trace the wrapped
+        function with the given arguments and stage it out to a Jaxpr. Then translate it to SDFG.
+        The result is encapsulated inside a `JaCeLowered` object which can later be compiled.
+
+        Note:
+            The call to the function is cached. As key an abstract description of the call,
+            similar to the tracers used by Jax, is used.
+            The tracing is always done with activated `x64` mode.
+        """
+        if len(kwargs) != 0:
+            raise NotImplementedError("Currently only positional arguments are supported.")
+
+        # TODO(phimuell): Currently the SDFG that we build only supports `C_CONTIGUOUS` memory
+        #  order. Since we support the paradigm that "everything passed to `lower()` should also
+        #  be accepted as argument to call the result", we forbid other memory orders here.
+        if not all((not util.is_array(arg)) or arg.flags["C_CONTIGUOUS"] for arg in args):
+            raise NotImplementedError("Currently can not yet handle strides beside 'C_CONTIGUOUS'.")
+
+        # In Jax `float32` is the main datatype, and they go to great lengths to avoid some
+        #  aggressive [type promotion](https://jax.readthedocs.io/en/latest/type_promotion.html).
+        #  However, in this case we will have problems when we call the SDFG, for some reasons
+        #  `CompiledSDFG` does not work in that case correctly, thus we enable it for the tracing.
+        with _jax.experimental.enable_x64():
```

Comment:
In an earlier version there was a warning here, but it become so annoying.
Also disabling is not really useful, so I keep it this way, because it the most useful and practical version.